core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.schema;
+
+import java.util.List;
+import java.util.stream.IntStream;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.UpdateSchema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Visitor class that accumulates the set of changes needed to evolve an existing schema into the union of the
+ * existing and a new schema. Changes are added to an {@link UpdateSchema} operation.
+ */
+public class UnionByNameVisitor extends SchemaWithPartnerVisitor<Integer, Boolean> {
+
+  private final UpdateSchema api;
+  private final Schema partnerSchema;
+
+  private UnionByNameVisitor(UpdateSchema api, Schema partnerSchema) {
+    this.api = api;
+    this.partnerSchema = partnerSchema;
+  }
+
+  /**
+   * Adds changes needed to produce a union of two schemas to an {@link UpdateSchema} operation.
+   * <p>
+   * Changes are accumulated to evolve the existingSchema into a union with newSchema.
+   *
+   * @param api an UpdateSchema for adding changes
+   * @param existingSchema an existing schema
+   * @param newSchema a new schema to compare with the existing
+   */
+  public static void visit(UpdateSchema api, Schema existingSchema, Schema newSchema) {
+    visit(newSchema, -1, new UnionByNameVisitor(api, existingSchema), new PartnerIdByNameAccessors(existingSchema));
+  }
+
+  @Override
+  public Boolean struct(Types.StructType struct, Integer partnerId, List<Boolean> missingPositions) {
+    if (partnerId == null) {
+      return true;
+    }
+
+    List<Types.NestedField> fields = struct.fields();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;struct&#39; (index: 0 | type: StructType). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;partnerId&#39;, &#39;missingPositions&#39;. The same parameter type is validated here for example: api/src/main/java/org/apache/iceberg/types/AssignFreshIds.java:78. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).